### PR TITLE
Block Support: Update border support UI

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Part of: https://github.com/WordPress/gutenberg/issues/31337
Related: https://github.com/WordPress/gutenberg/pull/32080
~Depends on: #31483 & https://github.com/WordPress/gutenberg/pull/31641~

## Description

This PR refines the border block support UI in line with https://github.com/WordPress/gutenberg/issues/31337. 

These refinements are separate to migrating the border block support to use the new progressive display panel proposed in https://github.com/WordPress/gutenberg/pull/32392. Converting to this new panel will be best handled separately.

The condensed color control proposed in the G2 components still requires some additional dependencies being imported into Gutenberg so will also be addressed in a future PR as soon as the control is available.

## How has this been tested?
- Manually in the editor. 
- Updated phpunit test for stylesheet generation

#### Test instructions
1. Checkout this PR branch and build
2. Enable border support UI via theme.json - [example theme.json gist](https://gist.github.com/aaronrobertshaw/8ed108e4a024b81297a55fc498ef3a57)
3. Add group block to a post, select it
3. Confirm border support controls display and function as expected

## Screenshots 
UpdatedBorderSupportUI

## Types of changes
New feature. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
The `dashed`, `dotted` and `solid` border style icons have been added to the `wordpress/icons` package now.

**Comment by jasmussen** [comment]
Sorry for missing this one. Based on the GIF, this is a great improvement. Thanks for working on this. 

The ability to have per-corner custom radii is a very compelling feature.

**Comment by aaronrobertshaw** [comment]
> Sorry for missing this one. Based on the GIF, this is a great improvement. Thanks for working on this.

Not a problem at all.

I deliberately left this as only a WIP draft as I wasn't sure whether the two PRs it depends on would force any changes. It seemed a little premature to flag it for full review until the custom units and per corner radii changes were merged. Hopefully those will be ready to merge in the near future.

**Comment by ramonjd** [comment]
Loving this so far. Editor and frontend look as expected. Inline styles on the frontend checkout.

May-28-2021 16-47-37

Screen Shot 2021-05-28 at 4 41 19 pm

**Comment by aaronrobertshaw** [comment]
I've rebased and re-tested this PR after recent changes to https://github.com/WordPress/gutenberg/pull/31641. It still works for me.

**Comment by aaronrobertshaw** [comment]
@jasmussen The PRs this one depends on, for custom CSS units and per-corner radii, have been merged. I think it would be good to get these refinements merged fairly soon if possible. Do you have any thoughts on this?

There are a couple of further refinements planned. They are:
- adding the condensed color control once it's available
- using the [proposed progressive discovery panel](#32392) once it lands
- setting the border style to none when border width is set to `0` ([PR](https://github.com/WordPress/gutenberg/pull/32080) already open for this)

I don't think these should be blockers for the refinements made in this PR. Merging this one will also make the above improvements easier and allow them to start sooner. If you are happy with these refinements so far, we'll just need a code review to proceed.

**Comment by jasmussen** [comment]
I share your general assessment that we now have the tools available to refine the prominence of these tools in relativelly short order using the progressive discovery panel and other improvements you've made. Which, by the way, is a major milestone: thank you for your patience and work on this, it will soon enable a ton of nice features. 

In that vein, I'd be happy to land this one as a step on that path. The only concern I still have is whether we can find something better than a dedicated "Border" panel; to me, the border is a property of the apperance of the block, just as background and perhaps other colors as well. But it seems okay to go with the border panel for now, since we'll likely need a collapsing color swatch component before we can make significant progress on it.

So get this one a code review and let's move to the next steps. Thanks again!

**Comment by aaronrobertshaw** [comment]
Great! Thanks for the quick reply 👍 

> The only concern I still have is whether we can find something better than a dedicated "Border" panel;

This is something else that I should have added to the future improvements list. Given the border panel already exists, improving it even in the short term is a win. 

I have some ideas regarding a potential "Appearance" panel wrapping together several block support provided panels or more specifically their controls. Those will best be explored in a separate PR but will benefit from having these more refined controls already in place.

> So get this one a code review and let's move to the next steps.

I've just realised this PR will need a couple of code tweaks to bring it inline with improvements made recently around providing custom CSS units for UnitControls. I'll make those changes tomorrow, then begin getting this reviewed.

Thanks as always for your help and guidance 🙇

**Comment by andrewserong** [comment]
@aaronrobertshaw this is testing well for me with the Group block (and nice improvements overall to the UI — I particularly like the border style buttons). I noticed that for some reason the individual unlinked radii isn't working for me on the Button block: the individual sides don't appear to be updating the inline style:

border-control-testing

In the above screenshot, this is running in TwentyTwenty theme after switching over from TT1-blocks (to test that the individual radii is persisted without issue after a theme change 👍, however I couldn't get the individual radii to work in the Button block on TT1-blocks or TwentyTwenty). I couldn't see anything obvious about why it mightn't be updating in the `onChange` handlers, but thought you might have a better idea of what's going on?

**Comment by aaronrobertshaw** [comment]
> I noticed that for some reason the individual unlinked radii isn't working for me on the Button block: the individual sides don't appear to be updating the inline style:

Nice catch 👍 

This is due to the button block not relying on the block support to generate the appropriate classes and styles. The button block needs updating manually to update its ad hoc application of the border radius. I've created a PR to solve this in https://github.com/WordPress/gutenberg/pull/33017.

There is also the search block that applies border radii from the block support in an ad hoc manner. That one is a little more complicated as adjustments are made to the border radius value before applying to inner elements as well as the wrapper. I'll address that in a separate PR. 

It might be that the border radius support itself needs a tweak to allow a block to only opt in for the shorthand border radius style version. That would then also mean needing to change this border radius control to exclude the "unlink" button in that case.

**Comment by aaronrobertshaw** [comment]
A PR is also up to make the search block's ad hoc application of the border radius block support styles compatible with the recent per-corner border radius changes. https://github.com/WordPress/gutenberg/pull/33023

The search and button blocks were the only blocks I found that opted into the border radius support and applied that support's styles manually. 

Ideally, we can land https://github.com/WordPress/gutenberg/pull/33017 and https://github.com/WordPress/gutenberg/pull/33023 soon and merge this one not long after.

File: packages/block-editor/src/components/border-radius-control/all-input-control.js
Comment (andrewserong) [inline packages/block-editor/src/components/border-radius-control/all-input-control.js]: Do we need to set this to a default of `noop`? It looks like we're only passing this prop down to `UnitControl` which also has its default of `noop`, so we might be able to drop it here and remove the lodash import?

File: packages/block-editor/src/components/border-radius-control/index.js
Comment (andrewserong) [inline packages/block-editor/src/components/border-radius-control/index.js]: Default unit step values for each of the units were added to the UnitControl component in https://github.com/WordPress/gutenberg/pull/32692. I think this means we might be able to skip defining a step value in this component, and let UnitControl use its defaults?

File: packages/block-editor/src/components/border-radius-control/input-controls.js
Comment (andrewserong) [inline packages/block-editor/src/components/border-radius-control/input-controls.js:24]: Could we avoid importing `noop` and bail early in this function if `! onChange`?

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (andrewserong) [inline packages/block-editor/src/components/border-radius-control/utils.js]: I know this one's copied from the existing box controls utils, but I'm wondering if `isNumber` from lodash is needed? Would `value === 0 || value` suffice for the check, given that (I think) we'd want all other falsy values to result in `null`?

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (andrewserong) [inline packages/block-editor/src/components/border-radius-control/utils.js]: Tiny nit: would this be more readable as `hasMixedValues` (though I assume there's a precedent for this naming elsewhere in Gutenberg?)

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (andrewserong) [inline packages/block-editor/src/components/border-radius-control/utils.js]: Tiny nit, similar to above: would this be more readable as `hasDefinedValues` (though I assume there's a precedent for this naming elsewhere in Gutenberg?)

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (andrewserong) [inline packages/block-editor/src/components/border-radius-control/utils.js]: Since the values object is one with known properties (we're dealing with border radius instead of a general purpose `isValuesDefined` function), can we do a more explicit check for the props we're looking for, and remove the lodash `isEmpty` import?

File: packages/edit-site/src/components/sidebar/border-panel.js
Comment (andrewserong) [inline packages/edit-site/src/components/sidebar/border-panel.js]: Similar comment to above, since the UnitControl component now has its own step defaults, do we still need the step logic in this component?

**Review by andrewserong — COMMENTED** [review]
This is working nicely for me @aaronrobertshaw, just gave it another test. I left a bunch of comments, mostly addressing things that were brought up by other folks when they were reviewing my BoxControl PR (https://github.com/WordPress/gutenberg/pull/32610), for example, seeing if we can get away with removing lodash imports where we can.

There's a couple of failing e2e tests, that I assume might be fixed up with a rebase (and getting the fixes for the Button and Search block rolled into this PR, of course).

Nice work again, it's very cool being able to adjust the individual radii, and the slimline border style looks really nice!

File: packages/block-editor/src/components/border-radius-control/index.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/index.js]: This was originally add prior to the change in #32692. The one reason we might wish to keep it here is for the `RangeControl` used in conjunction with the `UnitControl` fields. I'll see if we can use the new default step values contained within the `UnitControl` for the `RangeControl`.

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/utils.js]: This naming followed the functions within the `BoxControl`. I'm happy to improve the naming to use your suggestion though.

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/utils.js]: Same as with `isValuesMixed`, this followed the functions in `BoxControl`. I can update the function to the improved suggestion.

File: packages/block-editor/src/components/border-radius-control/all-input-control.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/all-input-control.js]: ✅ Done.

File: packages/block-editor/src/components/border-radius-control/input-controls.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/input-controls.js:24]: ✅ Done.

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/utils.js]: ✅ Done. I've also replaced the use of `isObject` to be able to completely remove the lodash import in `utils.js`

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/utils.js]: ✅ Done.

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/utils.js]: ✅ Done.

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/utils.js]: ✅ Done.

File: packages/edit-site/src/components/sidebar/border-panel.js
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar/border-panel.js]: ✅ Done.

**Comment by aaronrobertshaw** [comment]
Thanks @andrewserong for the detailed review 👍 

I've made the refinements, removed the use of lodash and also tweaked the use of unit steps for both `UnitControl` and `RangeControl` fields.

> There's a couple of failing e2e tests, that I assume might be fixed up with a rebase (and getting the fixes for the Button and Search block rolled into this PR, of course).

With these change in place, I'll rebase shortly and then address any e2e failures. The Button and Search block fixes have already been merged into trunk.

File: packages/block-editor/src/components/border-radius-control/index.js
Comment (ramonjd) [inline packages/block-editor/src/components/border-radius-control/index.js:29]: Any reason for the `null` value here? Is it the same as assigning `0`? If so, could we keep the same type for consistency?

**Review by glendaviesnz — APPROVED** [review]
This tested well for me in editor and frontend, and the UI works well switching between multiple and single border settings and retaining mixed values in these instances, etc.

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (ramonjd) [inline packages/block-editor/src/components/border-radius-control/utils.js]:
```suggestion
 * Gets the item with the most occurrence within an array
```

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (ramonjd) [inline packages/block-editor/src/components/border-radius-control/utils.js]: Feel free to ignore. I just noticed all the other return doc descriptions were indented.

**Review by ramonjd — COMMENTED** [review]
I took the UI for a run in Chrome/FF/Safari, and the controls work and display as expected in the block editor and frontend 👍 

I don't know if this PR should be concerned or whether it's a theme responsibility, but I checked in a couple of themes, and in Twenty Nineteen at least, the theme styles are overriding the primary and secondary border colors (e.g., `.entry .entry-content .has-primary-background-color`):

**Editor** 
Screen Shot 2021-07-07 at 9 59 22 am


**Frontend**
Screen Shot 2021-07-07 at 9 59 25 am

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/utils.js]: ✅ Done.

File: packages/block-editor/src/components/border-radius-control/utils.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/utils.js]: ✅ Done.

File: packages/block-editor/src/components/border-radius-control/index.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/index.js:29]: Good question. This is deliberately `null`. This allows the split border radius controls to remain empty as `0` is otherwise a valid value. It also means we can omit unnecessary longhand CSS styles and inherit values from the theme.json or global styles.

**Comment by aaronrobertshaw** [comment]
Thank you @glendaviesnz & @ramonjd for the reviews 👍 

> I don't know if this PR should be concerned or whether it's a theme responsibility, but I checked in a couple of themes, and in Twenty Nineteen at least, the theme styles are overriding the primary and secondary border colors 

This PR is primarily concerned with the UI around the existing border block support. My thinking is that it will be best to keep the border block support consistent with other block support features in terms of style specificity. If there is a particular reason the TwentyNineteen theme can't be updated to address its overwriting of border color styles, we can address it in a separate PR.


-----

I've made the requested tweaks as well as updated the border width control to use the new default step values for each unit. 

Let me know if I've missed anything.

**Review by ramonjd — APPROVED** [review]
Thanks for addressing those minor points, and clarifying the theme issue. 

LGTM 🚀

**Comment by aaronrobertshaw** [comment]
After merging this is working well for existing blocks that leverage border support.

https://user-images.githubusercontent.com/60436221/124694920-88fd0780-df25-11eb-816c-5f065d734dce.mp4

File: packages/block-editor/src/components/border-radius-control/index.js
Comment (oandregal) [inline packages/block-editor/src/components/border-radius-control/index.js:52]: Does this need to use the `spacing.units` or `layout.units` available via `theme.json`? I've noticed it's one of the few places where these aren't used so are not configurable by the theme.

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/index.js:52]: Good point.

I can see theme developers wishing to allow different units for borders to those of spacing or layouts as well. A PR allowing configuration of border units can be found in https://github.com/WordPress/gutenberg/pull/33315.

**Pull request merged**
